torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -195,35 +177,40 @@ void DistEngine::computeDependencies(
   autogradContext->setGraphTask(std::move(graphTask));
 }
 
-void DistEngine::runEngineAndAccumulateGradients(
+std::shared_ptr<FutureVariableList> DistEngine::runEngineAndAccumulateGradients(
     const ContextPtr& autogradContext,
     const std::shared_ptr<Node>& graphRoot,
     const edge_list& outputEdges) {
-  // Kick off autograd computation with the root node and retrieve all the
-  // gradients.
-  // TODO: make this non-blocking
-  // (https://github.com/pytorch/pytorch/issues/26359)
-  ThreadsBlockedGuard guard;
-  variable_list grads = engine_
-                            .execute_with_graph_task(
-                                autogradContext->retrieveGraphTask(), graphRoot)
-                            ->wait();
-
-  // Accumulate all the gradients in the context.
-  TORCH_INTERNAL_ASSERT(grads.size() == outputEdges.size());
-  for (size_t i = 0; i < grads.size(); i++) {
-    // It is possible that the grad is not defined since a separate invocation
-    // of the autograd engine on the same node might actually compute this
-    // gradient.
-    // Also accumulate grads only for AccumulateGrad function.
-    if (grads[i].defined() &&
-        dynamic_cast<AccumulateGrad*>(outputEdges[i].function.get())) {
-      auto& variable =
-          std::static_pointer_cast<AccumulateGrad>(outputEdges[i].function)
-              ->variable;
-      autogradContext->accumulateGrad(variable, grads[i]);
-    }
-  }
+  auto futureGrads = engine_.execute_with_graph_task(
+      autogradContext->retrieveGraphTask(), graphRoot);
+
+  futureGrads->addCallback(
+      [autogradContext, outputEdges](
+          const variable_list& grads,
+          const c10::optional<torch::utils::FutureError>& error) {
+        if (error) {
+          // Don't accumulate gradients if we receive an error.
+          return;
+        }
+
+        // Accumulate all the gradients in the context.
+        TORCH_INTERNAL_ASSERT(grads.size() == outputEdges.size());
+        for (size_t i = 0; i < grads.size(); i++) {
+          // It is possible that the grad is not defined since a separate
+          // invocation of the autograd engine on the same node might actually
+          // compute this gradient. Also accumulate grads only for
+          // AccumulateGrad function.
+          if (grads[i].defined() &&
+              dynamic_cast<AccumulateGrad*>(outputEdges[i].function.get())) {
```

Comment:
doesn't need to be solved in this change: When I see the dynamic_cast<>, I wonder a bit about how we handling throwing exceptions in our Future callbacks.

Right now, we're basically ignoring them, which probably mostly works. Not sure what's better here (there's not really a good way to change the existing completed future to an error), but it might be worth thinking about...

Reply:
Agree that this case is going to be completely benign.